dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/GitHubSourceCodeRepo.java

```diff
@@ -1055,27 +1055,92 @@ public String getRepositoryId(Entry entry) {
         }
     }
 
+
+
     @Override
     public String getMainBranch(Entry entry, String repositoryId) {
         String mainBranch = null;
 
         // Get repository based on username and repo id
+        mainBranch = getDefaultBranch(repositoryId);
+        if (mainBranch == null) {
+            return null;
+        }
+        // Determine which branch to use for tool info
+        if (entry.getDefaultVersion() != null) {
+            mainBranch = getBranchNameFromDefaultVersion(entry);
+        }
+
+        return mainBranch;
+    }
+
+    @Override
+    public String getDefaultBranch(String repositoryId) {
         if (repositoryId != null) {
             try {
                 GHRepository repository = github.getRepository(repositoryId);
-                // Determine the default branch on Github
-                mainBranch = repository.getDefaultBranch();
+                // Determine the default branch on GitHub
+                return repository.getDefaultBranch();
             } catch (IOException e) {
                 LOG.error("Unable to retrieve default branch for repository " + repositoryId, e);
                 return null;
             }
         }
-        // Determine which branch to use for tool info
-        if (entry.getDefaultVersion() != null) {
-            mainBranch = getBranchNameFromDefaultVersion(entry);
+        return null;
+    }
+
+    /**
+     * Detect branches that include a .dockstore.yml using a variety of heuristics.
+     * <p>
+     * Heuristics include looking at the default branch, commits that just touched a .dockstore.yml at the end of a branch, etc. while trying to avoid too many
+     * API calls, particularly an unbounded number of calls based on factors that can change between repos (number of tags, branches, etc.)
+     *
+     * @param repositoryId
+     * @return
+     */
+    public Set<String> detectDockstoreYml(String repositoryId) {
+        Set<String> candidateBranches = new HashSet<>();
+        if (repositoryId != null) {
+            try {
+                GHRepository repository = getRepository(repositoryId);
+                // see if there is a .dockstore.yml on any path that was just added to a branch/the HEAD
+                // unfortunately, there is no nice way to map this to a specific tag or branch, arbitrarily pick 2 to try
+                final int arbitraryNumberOfGuesses = 2;
+                final List<GHCommit> ghCommits = new ArrayList<>(repository.queryCommits().path(DOCKSTORE_YML_PATH).pageSize(arbitraryNumberOfGuesses).list().toList());
+                ghCommits.addAll(repository.queryCommits().path(DOCKSTORE_ALTERNATE_YML_PATH).pageSize(arbitraryNumberOfGuesses).list().toList());
```

Comment:
Do each of these `queryCommits...list()` calls return the latest two, or a random two?  If the latter, and you can increase the `pageSize` without incurring more calls, and there's some date information returned (either here or in the later `listBranchesWhereHead`), you could sort the list by date and take the latest two.  Although what you've got here probably works well for most real world cases...

Reply:
Unknown, I'm assuming random two, the frustrating thing is that many of these calls (list commits, list branches, etc.) paginate but do not have sorting parameters.